### PR TITLE
fix: #2470 Use mirror to download dependencies if applicable

### LINKED ISSUE #2470: Quarkus bypasses Nexus and downloads artefacts directly from maven central
Good evening, 

trying to work with Quarkus 0.15.0 from the office this afternoon, I realised that even though my Maven installation had our Nexus repository configured, Quarkus tried to download artefacts directly from Maven central, which didn't work as such a connection is prohibited at the office.
I thought that it was a misconfiguration of my Maven but working from home, also with Nexus as repository installed, the same behaviour happened and I can see if by taking a look at the output:
```
Downloading from nexus: http://localhost:8081/repository/maven-public/io/quarkus/quarkus-hibernate-orm-panache-deployment/0.15.0/quarkus-hibernate-orm-panache-deployment-0.15.0.jar
Downloaded from nexus: http://localhost:8081/repository/maven-public/io/quarkus/quarkus-narayana-jta-deployment/0.15.0/quarkus-narayana-jta-deployment-0.15.0.jar (7.7 kB at 3.4 kB/s)
Downloaded from nexus: http://localhost:8081/repository/maven-public/io/quarkus/quarkus-agroal-deployment/0.15.0/quarkus-agroal-deployment-0.15.0.jar (11 kB at 4.8 kB/s)
Downloaded from nexus: http://localhost:8081/repository/maven-public/io/quarkus/quarkus-hibernate-orm-deployment/0.15.0/quarkus-hibernate-orm-deployment-0.15.0.jar (48 kB at 19 kB/s)
Downloaded from nexus: http://localhost:8081/repository/maven-public/io/quarkus/quarkus-hibernate-orm-panache-deployment/0.15.0/quarkus-hibernate-orm-panache-deployment-0.15.0.jar (30 kB at 11 kB/s)
Downloaded from nexus: http://localhost:8081/repository/maven-public/io/quarkus/quarkus-smallrye-opentracing-deployment/0.15.0/quarkus-smallrye-opentracing-deployment-0.15.0.jar (6.2 kB at 1.5 kB/s)
[INFO] [io.quarkus.deployment.QuarkusAugmentor] Beginning quarkus augmentation
[INFO] [org.jboss.threads] JBoss Threads version 3.0.0.Beta3
[WARNING] [io.quarkus.agroal.deployment.AgroalProcessor] Agroal dependency is present but no driver has been defined for the default datasource
[INFO] [org.hibernate.jpa.boot.internal.PersistenceXmlParser] HHH000318: Could not find any META-INF/persistence.xml file in the classpath
[INFO] [org.hibernate.Version] HHH000412: Hibernate Core {5.4.2.Final}
[INFO] [io.quarkus.deployment.QuarkusAugmentor] Quarkus augmentation completed in 4147ms
[INFO] [io.quarkus.creator.phase.runnerjar.RunnerJarPhase] Building jar: /Volumes/Work/projets/immobesichtigung24/ib24-modal/target/ib24-modal-runner.jar
[INFO] 
[INFO] --- quarkus-maven-plugin:0.15.0:dev (default-cli) @ ib24-modal ---
Downloading from central: https://repo.maven.apache.org/maven2/io/quarkus/quarkus-resteasy/0.15.0/quarkus-resteasy-0.15.0.jar
Downloading from central: https://repo.maven.apache.org/maven2/io/quarkus/quarkus-smallrye-health/0.15.0/quarkus-smallrye-health-0.15.0.jar
Downloading from central: https://repo.maven.apache.org/maven2/io/quarkus/quarkus-smallrye-opentracing/0.15.0/quarkus-smallrye-opentracing-0.15.0.jar
Downloading from central: https://repo.maven.apache.org/maven2/io/quarkus/quarkus-smallrye-metrics/0.15.0/quarkus-smallrye-metrics-0.15.0.jar
Downloading from central: https://repo.maven.apache.org/maven2/io/quarkus/quarkus-resteasy-jsonb/0.15.0/quarkus-resteasy-jsonb-0.15.0.jar
Downloaded from central: https://repo.maven.apache.org/maven2/io/quarkus/quarkus-smallrye-opentracing/0.15.0/quarkus-smallrye-opentracing-0.15.0.jar (0 B at 0 B/s)
Downloading from central: https://repo.maven.apache.org/maven2/io/quarkus/quarkus-hibernate-orm-panache/0.15.0/quarkus-hibernate-orm-panache-0.15.0.jar
Downloaded from central: https://repo.maven.apache.org/maven2/io/quarkus/quarkus-smallrye-metrics/0.15.0/quarkus-smallrye-metrics-0.15.0.jar (0 B at 0 B/s)
Downloaded from central: https://repo.maven.apache.org/maven2/io/quarkus/quarkus-resteasy-jsonb/0.15.0/quarkus-resteasy-jsonb-0.15.0.jar (0 B at 0 B/s)
Downloaded from central: https://repo.maven.apache.org/maven2/io/quarkus/quarkus-smallrye-health/0.15.0/quarkus-smallrye-health-0.15.0.jar (0 B at 0 B/s)
Downloaded from central: https://repo.maven.apache.org/maven2/io/quarkus/quarkus-resteasy/0.15.0/quarkus-resteasy-0.15.0.jar (0 B at 0 B/s)
Downloaded from central: https://repo.maven.apache.org/maven2/io/quarkus/quarkus-hibernate-orm-panache/0.15.0/quarkus-hibernate-orm-panache-0.15.0.jar (0 B at 0 B/s)
Downloading from central: https://repo.maven.apache.org/maven2/io/quarkus/quarkus-resteasy-deployment/0.15.0/quarkus-resteasy-deployment-0.15.0.jar
Downloading from central: https://repo.maven.apache.org/maven2/io/quarkus/quarkus-smallrye-health-deployment/0.15.0/quarkus-smallrye-health-deployment-0.15.0.jar
Downloading from central: https://repo.maven.apache.org/maven2/io/quarkus/quarkus-smallrye-opentracing-deployment/0.15.0/quarkus-smallrye-opentracing-deployment-0.15.0.jar
Downloading from central: https://repo.maven.apache.org/maven2/io/quarkus/quarkus-smallrye-metrics-deployment/0.15.0/quarkus-smallrye-metrics-deployment-0.15.0.jar
Downloading from central: https://repo.maven.apache.org/maven2/io/quarkus/quarkus-resteasy-jsonb-deployment/0.15.0/quarkus-resteasy-jsonb-deployment-0.15.0.jar
Downloaded from central: https://repo.maven.apache.org/maven2/io/quarkus/quarkus-smallrye-health-deployment/0.15.0/quarkus-smallrye-health-deployment-0.15.0.jar (0 B at 0 B/s)
Downloaded from central: https://repo.maven.apache.org/maven2/io/quarkus/quarkus-resteasy-deployment/0.15.0/quarkus-resteasy-deployment-0.15.0.jar (0 B at 0 B/s)
Downloaded from central: https://repo.maven.apache.org/maven2/io/quarkus/quarkus-resteasy-jsonb-deployment/0.15.0/quarkus-resteasy-jsonb-deployment-0.15.0.jar (0 B at 0 B/s)
Downloading from central: https://repo.maven.apache.org/maven2/io/quarkus/quarkus-hibernate-orm-panache-deployment/0.15.0/quarkus-hibernate-orm-panache-deployment-0.15.0.jar
Downloaded from central: https://repo.maven.apache.org/maven2/io/quarkus/quarkus-smallrye-metrics-deployment/0.15.0/quarkus-smallrye-metrics-deployment-0.15.0.jar (0 B at 0 B/s)
Downloaded from central: https://repo.maven.apache.org/maven2/io/quarkus/quarkus-smallrye-opentracing-deployment/0.15.0/quarkus-smallrye-opentracing-deployment-0.15.0.jar (0 B at 0 B/s)
Downloaded from central: https://repo.maven.apache.org/maven2/io/quarkus/quarkus-hibernate-orm-panache-deployment/0.15.0/quarkus-hibernate-orm-panache-deployment-0.15.0.jar (0 B at 0 B/s)
Listening for transport dt_socket at address: 5005
```
One can see that "quarkus:dev" seems to be the culprit. 
Is my analysis of the problem correct? If yes, is there something to do to prevent this?

Regards, 
Daniel W.

### PR DESCRIPTION
**Pull request by jycr**
fix: #2470 Use mirror to download dependencies if applicable.
And ensure that proxy config is setup if needed

### CODE REVIEW

**Comment by aloubyansky** [comment]
@jycr wouldn't this change https://github.com/aloubyansky/quarkus/commit/e5a1bac287171e81e6d0ab9ebc583ff461ca7ee5 be enough?

**Comment by jycr** [comment]
> @jycr wouldn't this change [aloubyansky@e5a1bac](https://github.com/aloubyansky/quarkus/commit/e5a1bac287171e81e6d0ab9ebc583ff461ca7ee5) be enough?
I think there still little problems with your proposal:
* if a mirror is applicable for several RemoteRepository, we will have several same RemoteRepository definition in return result array
* if a mirror need to be connected through proxy, it doesn't work because proxy setup, according mirror target, is not applied

**Comment by jycr** [comment]
I hope It  is not too late for this PR for  0.16.0 release ?

File: independent-projects/bootstrap/core/src/main/java/io/quarkus/bootstrap/resolver/maven/ProxyAwareMirrorSelector.java [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+    }
+
+    @Override
+    public RemoteRepository getMirror(final RemoteRepository remoteRepository) {
+        final RemoteRepository remoteRepositoryMirror = ofNullable(wrappedMirrorSelector)
+                .map(ms -> ms.getMirror(remoteRepository))
+                .orElse(null);
+        if (remoteRepositoryMirror == null) {
+            // No Mirror, return remoteRepository it-self
+            return remoteRepository;
```
Comment (aloubyansky) [inline independent-projects/bootstrap/core/src/main/java/io/quarkus/bootstrap/resolver/maven/ProxyAwareMirrorSelector.java]: Shouldn't it be checked whether a proxy is relevant for it?

File: independent-projects/bootstrap/core/src/main/java/io/quarkus/bootstrap/resolver/maven/ProxyAwareMirrorSelector.java [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+    ProxyAwareMirrorSelector(List<Mirror> mirrors, ProxySelector proxySelector) {
+        this.wrappedMirrorSelector = createMirrorSelector(mirrors);
+        this.proxySelector = proxySelector;
+    }
+
+    @Override
+    public RemoteRepository getMirror(final RemoteRepository remoteRepository) {
+        final RemoteRepository remoteRepositoryMirror = ofNullable(wrappedMirrorSelector)
+                .map(ms -> ms.getMirror(remoteRepository))
+                .orElse(null);
```
Comment (aloubyansky) [inline independent-projects/bootstrap/core/src/main/java/io/quarkus/bootstrap/resolver/maven/ProxyAwareMirrorSelector.java]: Please change it to `wrappedMirrorSelector == null ? null : wrappedMirrorSelector.getMirror(remoteRepository)`

File: independent-projects/bootstrap/core/src/main/java/io/quarkus/bootstrap/resolver/maven/ProxyAwareMirrorSelector.java [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+        if (remoteRepositoryMirror == null) {
+            // No Mirror, return remoteRepository it-self
+            return remoteRepository;
+        }
+
+        final Proxy proxy = ofNullable(proxySelector)
+                .map(ps -> ps.getProxy(remoteRepositoryMirror))
+                .orElse(null);
```
Comment (aloubyansky) [inline independent-projects/bootstrap/core/src/main/java/io/quarkus/bootstrap/resolver/maven/ProxyAwareMirrorSelector.java]: Please change it to `proxySelector == null ? null : proxySelector.getProxy(remoteRepositoryMirror);`

File: independent-projects/bootstrap/core/src/main/java/io/quarkus/bootstrap/resolver/maven/MavenRepoInitializer.java [hunk 4]
```diff
@@ -295,11 +304,65 @@ public void add(ModelProblemCollectorRequest req) {
...
+                                                        .addUsername(sp.getUsername())
+                                                        .addPassword(sp.getPassword())
+                                                        .build()
+                                        ))
+                                .map(proxy -> (ProxySelector) new DefaultProxySelector().add(
+                                        proxy,
+                                        settingsProxy.getNonProxyHosts()
+                                ))
+                                .orElse(NULL_PROXY_SELECTOR))
+                .orElse(NULL_PROXY_SELECTOR);
```
Comment (aloubyansky) [inline independent-projects/bootstrap/core/src/main/java/io/quarkus/bootstrap/resolver/maven/MavenRepoInitializer.java]: Could you please avoid lambdas here?

**Review by aloubyansky — CHANGES_REQUESTED** [review]
No, it's not too late. I've added a few remarks, could you please address those? Please avoid lambdas when they are not necessary. Thanks!

File: independent-projects/bootstrap/core/src/main/java/io/quarkus/bootstrap/resolver/maven/ProxyAwareMirrorSelector.java (hunk 1 above)
Comment (jycr) [reply independent-projects/bootstrap/core/src/main/java/io/quarkus/bootstrap/resolver/maven/ProxyAwareMirrorSelector.java]: You right. I will fix it

**Comment by jycr** [comment]
I have simplified implementation and move proxy setup (and NULL_PROXY_SELECTOR) into ProxyAwareMirrorSelector

File: independent-projects/bootstrap/core/src/test/java/io/quarkus/bootstrap/resolver/maven/MavenRepoInitializerTest.java [hunk 5]
```diff
@@ -0,0 +1,138 @@
+package io.quarkus.bootstrap.resolver.maven;
```
Comment (aloubyansky) [inline independent-projects/bootstrap/core/src/test/java/io/quarkus/bootstrap/resolver/maven/MavenRepoInitializerTest.java]: Just to be consistent with the rest of the tests in this module, please use `io.quarkus.bootstrap.resolver.maven.test` as a package.

File: independent-projects/bootstrap/core/src/test/java/io/quarkus/bootstrap/resolver/maven/MavenRepoInitializerTest.java [hunk 6]
```diff
@@ -0,0 +1,138 @@
...
+        localProxy.setId("local-proxy-http");
+
+        mirrorA = new Mirror();
+        mirrorA.setMirrorOf("central,jboss-public-repository,spring-public-repository");
+        mirrorA.setUrl("https://mirror.com/artifact/content/groups/public/");
+        mirrorA.setId("mirror-A");
+    }
+
+    @Test
+    public void get_remote_repo_from_settings_without_proxy_neither_mmirror() throws AppModelResolverException {
```
Comment (aloubyansky) [inline independent-projects/bootstrap/core/src/test/java/io/quarkus/bootstrap/resolver/maven/MavenRepoInitializerTest.java]: We are using camelCase for method names all over, let's stick to that. Also let's rephrase it a bit, e.g. `getRemoteRepoFromSettingsWithNeitherProxyNorMirror`

File: independent-projects/bootstrap/core/src/test/java/io/quarkus/bootstrap/resolver/maven/MavenRepoInitializerTest.java [hunk 7]
```diff
@@ -0,0 +1,138 @@
...
+        assertTrue(repos.get(0).getMirroredRepositories().isEmpty());
+
+        final RemoteRepository centralRepo = repos.get(repos.size() - 1);
+        assertEquals("central", centralRepo.getId());
+        assertNull(centralRepo.getProxy());
+        assertTrue(centralRepo.getMirroredRepositories().isEmpty());
+    }
+
+    @Test
+    public void get_remote_repo_from_settings_with_proxy_without_mirror() throws AppModelResolverException {
```
Comment (aloubyansky) [inline independent-projects/bootstrap/core/src/test/java/io/quarkus/bootstrap/resolver/maven/MavenRepoInitializerTest.java]: camelCase please

File: independent-projects/bootstrap/core/src/test/java/io/quarkus/bootstrap/resolver/maven/MavenRepoInitializerTest.java [hunk 8]
```diff
@@ -0,0 +1,138 @@
...
+        assertNotNull(repos.get(0).getMirroredRepositories());
+
+        final RemoteRepository centralRepo = repos.get(repos.size() - 1);
+        assertEquals("central repo must be added as default repository", "central", centralRepo.getId());
+        assertNotNull(centralRepo.getProxy());
+        assertTrue(centralRepo.getMirroredRepositories().isEmpty());
+    }
+
+    @Test
+    public void get_remote_repo_from_settings_with_proxy_and_mirror() throws AppModelResolverException {
```
Comment (aloubyansky) [inline independent-projects/bootstrap/core/src/test/java/io/quarkus/bootstrap/resolver/maven/MavenRepoInitializerTest.java]: camelCase please

**Review by aloubyansky — CHANGES_REQUESTED** [review]
A few minor remarks. Thanks a lot, @jycr !

**Comment by gsmet** [comment]
@aloubyansky I let you drive this one home before the release? Thanks!

**Comment by aloubyansky** [comment]
Ok, given that we don't have time, let's merge it and I'll follow up with adjustments.

**Review by aloubyansky — APPROVED** [review]

**Comment by gsmet** [comment]
@aloubyansky we have time. See my email on the list. I'll release tomorrow or Friday.

**Comment by aloubyansky** [comment]
Yes, just saw it. Ok. I'll wait a bit for @jycr to respond.

**Comment by jycr** [comment]
> Yes, just saw it. Ok. I'll wait a bit for @jycr to respond.

@aloubyansky : I do not see what remains to be done.
Have I forgotten something?

**Comment by aloubyansky** [comment]
Oh, you've already pushed! Sorry, I missed it. Merging... Thanks!

**Comment by aloubyansky** [comment]
well, waiting for the CI first

**Comment by jycr** [comment]
> well, waiting for the CI first

The CI has just finished 🙂

**Pull request merged**